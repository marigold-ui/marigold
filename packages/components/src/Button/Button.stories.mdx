import { Meta, Story, Preview } from '@storybook/addon-docs/blocks';
import { Button } from './Button';

<Meta title="Components/Button" />

# Button-Component

## Description

With the Button component you can render a `button` element.
By adding the `themeSection` prop you can change to another themeSection. `button` is the default section.
The variant in your section can be added with the variant prop. E.g. `variant="primary.large"` or `variant="secondary.small"`.
Specific css style prop can also be added.

## Properties

| Property       | Type        | Default    |
| :------------- | :---------- | :--------- |
| `themeSection` | `string`    | `'button'` |
| `variant`      | `string`    |            |
| `css`          | `css props` |            |

## Import

```tsx
import { Button } from '@marigold/components';
```

## Usage

### Large

<Preview>
  <Story name="primary.large">
    <Button>Primary</Button>
  </Story>
  <Story name="secondary.large">
    <Button>Secondary</Button>
  </Story>
  <Story name="disabled.large">
    <Button>Disabled</Button>
  </Story>
  <Story name="ghost.large">
    <Button>Ghost</Button>
  </Story>
  <Story name="ghostDisabled.large">
    <Button>Ghost Disabled</Button>
  </Story>
</Preview>

### Large Hovered

<Preview>
  <Story name="primaryHovered.large">
    <Button>Primary</Button>
  </Story>
  <Story name="secondaryHovered.large">
    <Button>Secondary</Button>
  </Story>
  <Story name="ghostHovered.large">
    <Button>Ghost</Button>
  </Story>
</Preview>

### Small

<Preview>
  <Story name="primary.small">
    <Button>Primary</Button>
  </Story>
  <Story name="secondary.small">
    <Button>Secondary</Button>
  </Story>
  <Story name="disabled.small">
    <Button>Disabled</Button>
  </Story>
  <Story name="ghost.small">
    <Button>Ghost</Button>
  </Story>
  <Story name="ghostDisabled.small">
    <Button>Ghost Disabled</Button>
  </Story>
</Preview>
