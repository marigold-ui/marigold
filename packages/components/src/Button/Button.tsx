import React, { forwardRef, ReactNode } from 'react';
import { useButton } from '@react-aria/button';
import { useFocusRing } from '@react-aria/focus';
import { useHover } from '@react-aria/interactions';
import { mergeProps, useObjectRef } from '@react-aria/utils';
import { PressEvents } from '@react-types/shared';

import {
  Box,
  ThemeExtension,
  useComponentStyles,
  useStateProps,
} from '@marigold/system';
import { HtmlProps, PolymorphicComponent, PropsOf } from '@marigold/types';

// Theme Extension
// ---------------
export interface ButtonThemeExtension extends ThemeExtension<'Button'> {}

// Props
// ---------------
export interface ButtonOwnProps extends PressEvents, HtmlProps<'button'> {
  children?: ReactNode;
  variant?: string;
  size?: string;
  fullWidth?: boolean;
}

export interface ButtonProps extends PropsOf<typeof Button> {}

// Component
// ---------------
export const Button = forwardRef(
  (
    {
      as = 'button',
      children,
      variant,
      size,
      disabled,
      onClick,
      onPress,
      onPressStart,
      onPressEnd,
      onPressChange,
      onPressUp,
      fullWidth,
      ...props
    },
    ref
  ) => {
    const buttonRef = useObjectRef<HTMLButtonElement>(ref as any);
    const { hoverProps, isHovered } = useHover({ isDisabled: disabled });
    const { isFocusVisible, focusProps } = useFocusRing({
      autoFocus: props.autoFocus,
    });
    const { buttonProps, isPressed } = useButton(
      {
        /**
         * `react-aria` only expected `Element` but we casted
         * it to a `HTMLButtonElement` internally.
         */
        ...(props as any),
        onClick,
        onPress,
        onPressStart,
        onPressEnd,
        onPressChange,
        onPressUp,
        elementType: typeof as === 'string' ? as : 'span',
        isDisabled: disabled,
      },
      buttonRef
    );

    const styles = useComponentStyles('Button', { variant, size });
    const stateProps = useStateProps({
      active: isPressed,
      focusVisible: isFocusVisible,
      hover: isHovered,
    });

    return (
      <Box
        {...mergeProps(buttonProps, focusProps, hoverProps, props)}
        {...stateProps}
        as={as}
        ref={buttonRef}
        __baseCSS={{
          display: 'inline-flex',
          alignItems: 'center',
          justifyContent: 'center',
          gap: '0.5ch',
          cursor: disabled ? 'not-allowed' : 'pointer',
          width: fullWidth ? '100%' : undefined,
          '&:focus': {
            outline: 0,
          },
        }}
        css={styles}
      >
        {children}
      </Box>
    );
  }
) as PolymorphicComponent<'button', ButtonOwnProps>;
