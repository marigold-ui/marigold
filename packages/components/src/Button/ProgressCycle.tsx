import { ProgressBar } from 'react-aria-components';
import type RAC from 'react-aria-components';

interface ProgressCycleProps extends RAC.ProgressBarProps {
  size?: string;
}

export const ProgressCycle = ({ size, ...props }: ProgressCycleProps) => {
  let strokeWidth = 2;

  // SVG strokes are centered, so subtract half the stroke width from the radius to create an inner stroke.
  let radius = `calc(50% - ${strokeWidth / 2}px)`;

  return (
    <ProgressBar {...props}>
      <svg width={16} height={16} fill="none">
        <circle
          cx="50%"
          cy="50%"
          r={radius}
          strokeWidth={strokeWidth}
          className="stroke-border-selected"
        />
        <circle
          cx="50%"
          cy="50%"
          r={radius}
          strokeWidth={strokeWidth}
          // Normalize the path length to 100 so we can easily set stroke-dashoffset to a percentage.
          pathLength="100"
          // Add extra gap between dashes so 0% works in Chrome.
          strokeDasharray="100 200"
          strokeLinecap="round"
          strokeDashoffset={undefined}
          className="animate-progress-cycle origin-center -rotate-90 stroke-black/55"
        />
      </svg>
    </ProgressBar>
  );
};
