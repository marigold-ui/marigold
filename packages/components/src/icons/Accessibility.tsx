// https://lucide.dev/icons/accessibility
import { cn } from '@marigold/system';
import { IconProps } from './Icons.types';

export const Accessibility = ({ size = 24, className, ...rest }: IconProps) => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width={size}
      height={size}
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
      className={cn(
        'lucide lucide-accessibility-icon lucide-accessibility',
        className
      )}
      {...rest}
    >
      <circle cx="16" cy="4" r="1" />
      <path d="m18 19 1-7-6 1" />
      <path d="m5 8 3-3 5.5 3-2.36 3.5" />
      <path d="M4.24 14.5a5 5 0 0 0 6.88 6" />
      <path d="M13.76 17.5a5 5 0 0 0-6.88-6" />
    </svg>
  );
};
