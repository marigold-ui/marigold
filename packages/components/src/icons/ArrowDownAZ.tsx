// https://lucide.dev/icons/arrow-down-a-z
import { cn } from '@marigold/system';
import { IconProps } from './Icons.types';

export const ArrowDownAZ = ({ size = 24, className, ...rest }: IconProps) => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width={size}
      height={size}
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
      className={cn(
        'lucide lucide-arrow-down-az-icon lucide-arrow-down-a-z',
        className
      )}
      {...rest}
    >
      <path d="m3 16 4 4 4-4" />
      <path d="M7 20V4" />
      <path d="M20 8h-5" />
      <path d="M15 10V6.5a2.5 2.5 0 0 1 5 0V10" />
      <path d="M15 14h5l-5 6h5" />
    </svg>
  );
};
