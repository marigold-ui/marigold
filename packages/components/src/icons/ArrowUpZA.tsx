// https://lucide.dev/icons/arrow-up-z-a
import { cn } from '@marigold/system';
import { IconProps } from './Icons.types';

export const ArrowUpZA = ({ size = 24, className, ...rest }: IconProps) => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width={size}
      height={size}
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
      className={cn(
        'lucide lucide-arrow-up-za-icon lucide-arrow-up-z-a',
        className
      )}
      {...rest}
    >
      <path d="m3 8 4-4 4 4" />
      <path d="M7 4v16" />
      <path d="M15 4h5l-5 6h5" />
      <path d="M15 20v-3.5a2.5 2.5 0 0 1 5 0V20" />
      <path d="M20 18h-5" />
    </svg>
  );
};
