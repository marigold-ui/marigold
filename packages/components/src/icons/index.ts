export * from './ChevronUp';
export * from './ChevronDown';
export * from './ChevronRight';
export * from './ChevronLeft';

export * from './SortDown';
export * from './SortUp';

export * from './Accessible';
