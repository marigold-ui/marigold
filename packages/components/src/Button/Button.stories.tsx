/* eslint-disable react-hooks/rules-of-hooks */
import React, { useState } from 'react';
import { Facebook } from '@marigold/icons';
import { Button } from './Button';
import isChromatic from 'chromatic';

import type { Meta, StoryObj } from '@storybook/react';

const meta = {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    disabled: {
      control: {
        type: 'boolean',
      },
      description: 'Disable the button',
      defaultValue: false,
    },
    fullWidth: {
      control: {
        type: 'boolean',
      },
      description: 'Take availble width',
      defaultValue: false,
    },
    size: {
      control: {
        type: 'text',
      },
      description: 'Size of the button',
    },
    variant: {
      control: {
        type: 'text',
      },
      table: {
        type: { summary: 'string' },
        defaultValue: { summary: 'primary' },
      },
      description: 'Variant of the button',
    },
    children: {
      control: 'text',
      description: 'Children of the button',
      table: {
        type: { summary: 'string' },
        defaultValue: { summary: 'Click me!' },
      },
    },
  },
  args: {
    variant: 'primary',
    children: 'Click me!',
  },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  render: args => <Button {...args} />,
};

export const WithIcon: Story = {
  render: ({ children, ...args }) => (
    <Button {...args}>
      <Facebook />
      {children}
    </Button>
  ),
};

export const OnPress: Story = {
  render: args => <Button {...args} onPress={(e: any) => console.log(e)} />,
};

export const AsProp: Story = {
  render: (args: any) => (
    <Button as="a" href="https://reservix.net" {...args} />
  ),
};

export const FullWidth: Story = {
  render: args => <Button {...args} fullWidth />,
};

export const PassThroughProps: Story = {
  render: args => {
    const [isHovered, setHovered] = useState(false);
    return (
      <>
        <Button
          {...args}
          onMouseEnter={() => setHovered(true)}
          onMouseLeave={() => setHovered(false)}
        />
        <br />
        state: {isHovered ? 'hovered' : 'not hovered'}
      </>
    );
  },
};

WithIcon.parameters = {
  chromatic: { viewports: [320, 1200] },
  theme: isChromatic() ? 'b2b' : 'stacked',
};
