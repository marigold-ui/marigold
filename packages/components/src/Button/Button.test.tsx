import React from 'react';
import { render, screen } from '@testing-library/react';
import { ThemeProvider } from '@marigold/system';
import { Button } from './Button';
import { Facebook } from '@marigold/icons';

const theme = {
  button: {
    large: {
      p: '16px',
    },
    small: {
      p: '16px',
    },
    primary: {
      fontFamily: 'Inter',
    },
    secondary: {
      fontFamily: 'Arial',
    },
  },
};

test('supports default variant', () => {
  render(
    <ThemeProvider theme={theme}>
      <Button>button</Button>
    </ThemeProvider>
  );
  const button = screen.getByText(/button/);

  expect(button.parentElement).toHaveStyle(`font-family: Inter`);
});

test('supports default size', () => {
  render(
    <ThemeProvider theme={theme}>
      <Button>button</Button>
    </ThemeProvider>
  );
  const button = screen.getByText(/button/);

  expect(button.parentElement).toHaveStyle(`padding: 16px`);
});

test('accepts other variant than default', () => {
  render(
    <ThemeProvider theme={theme}>
      <Button variant="secondary">button</Button>
    </ThemeProvider>
  );
  const button = screen.getByText(/button/);

  expect(button.parentElement).toHaveStyle(`font-family: Arial`);
});

test('accepts other size than default', () => {
  render(
    <ThemeProvider theme={theme}>
      <Button size="small">button</Button>
    </ThemeProvider>
  );
  const button = screen.getByText(/button/);

  expect(button.parentElement).toHaveStyle(`padding: 16px`);
});

test('renders <button> element', () => {
  render(
    <ThemeProvider theme={theme}>
      <Button>button</Button>
    </ThemeProvider>
  );
  const button = screen.getByText(/button/);

  expect(button.parentElement instanceof HTMLButtonElement).toBeTruthy();
  expect(button instanceof HTMLSpanElement).toBeTruthy();
});

test('accepts other button components', () => {
  const CustomButton = React.forwardRef<
    HTMLSpanElement,
    { children?: React.ReactNode }
  >(() => <span>I am a Button!</span>);

  render(
    <ThemeProvider theme={theme}>
      <Button as={CustomButton}>Button</Button>
    </ThemeProvider>
  );

  const button = screen.getByText('I am a Button!');
  expect(button).toBeTruthy();
});

test('add icon in button works as expected', () => {
  render(
    <ThemeProvider theme={theme}>
      <Button>
        <Facebook fill="red" size={30} title="facebook" />
        iconbutton
      </Button>
    </ThemeProvider>
  );
  const button = screen.getByText(/iconbutton/);
  const icon = screen.getByTitle(/facebook/);

  expect(button instanceof HTMLSpanElement).toBeTruthy();
  expect(button).toHaveStyle('display: inline-flex');
  expect(button.firstChild instanceof SVGElement).toBeTruthy();
  expect(icon.getAttribute('fill')).toEqual('red');
  expect(icon.getAttribute('width')).toEqual('30');
});

test('accepts custom styles prop className', () => {
  render(
    <ThemeProvider theme={theme}>
      <Button className="custom-class-name" title="button">
        click
      </Button>
    </ThemeProvider>
  );
  const button = screen.getByTitle(/button/);

  expect(button.className).toMatch('custom-class-name');
});
