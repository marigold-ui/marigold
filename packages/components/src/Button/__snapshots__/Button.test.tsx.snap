// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supports disabled prop 1`] = `
<button
  class="css-1xjkecy"
  disabled=""
  type="button"
>
  button
</button>
`;
